src/storage/animationsSchema.ts

```diff
@@ -0,0 +1,946 @@
+
+const filePath = z
+	.string()
+	.regex(
+		/^\w[^"*:<>?\\|]+\.\w{3,5}$/,
+		'String must be a valid filepath. The following characters are unsafe for cross-platform filesystems: "*:<>?\\|',
+	);
+
+const sequencerDBEntry = z
+	.string()
+	.regex(/^[\w-]+(\.([\w-]+|\{\w+(,\w+)+\}))+$/, 'String must be a valid Sequencer database entry.');
+
+const vector2 = z
+	.object({
+		x: z
+			.number()
+			.refine(...nonZero)
+			.optional(),
+		y: z
+			.number()
+			.refine(...nonZero)
+			.optional(),
+	})
+	.strict()
+	.refine(...nonEmpty);
+
+const soundEffect = z
+	.object({
+		type: z.string().min(1),
+		intensity: z.number().positive(),
+	})
+	.strict();
+const soundData = z
+	.object({
+		file: sequencerDBEntry.or(filePath),
+		waitUntilFinished: z.number().optional(),
+		atLocation: z
+			.object({
+				cacheLocation: z.literal(true).optional(),
+				offset: vector2.optional(),
+				randomOffset: z.number().optional(),
+				gridUnits: z.literal(true).optional(),
+				local: z.literal(true).optional(),
+			})
+			.strict()
+			.optional(),
+		radius: z.number().positive().optional(),
+		volume: z.number().positive().optional(),
+		duration: z.number().positive().optional(),
+		constrainedByWalls: z.literal(true).optional(),
+		predicate: z
+			.array(predicate)
+			.min(1)
+			.refine(...uniqueItems)
+			.optional(),
+		default: z.literal(true).optional(),
+		delay: z
+			.number()
+			.refine(...nonZero)
+			.optional(),
+		muffledEffect: soundEffect.optional(),
+		baseEffect: soundEffect.optional(),
+	})
+	.strict();
+const soundConfig = soundData.or(
+	z
+		.array(soundData)
+		.min(1)
+		.refine(...uniqueItems),
+);
+
+const presetOptions = z
+	.object({
+		attachTo: z.literal(true).or(
+			z
+				.object({
+					align: z.string().optional(),
+					edge: z.string().optional(),
+					bindVisibility: z.literal(true).optional(),
+					bindAlpha: z.literal(true).optional(),
+					bindScale: z.literal(true).optional(),
+					bindElevation: z.literal(true).optional(),
+					followRotation: z.literal(true).optional(),
+					offset: vector2.optional(),
+					randomOffset: z.number().optional(),
+					gridUnits: z.literal(true).optional(),
+					local: z.literal(true).optional(),
+				})
+				.strict()
+				.refine(...nonEmpty)
+				.optional(),
+		),
+		atLocation: z.literal(true).or(
+			z
+				.object({
+					cacheLocation: z.literal(true).optional(),
+					offset: z
+						.object({
+							x: z
+								.number()
+								.refine(...nonZero)
+								.or(z.tuple([z.number().refine(...nonZero), z.number().refine(...nonZero)]))
+								.optional(),
+							y: z
+								.number()
+								.refine(...nonZero)
+								.or(z.tuple([z.number().refine(...nonZero), z.number().refine(...nonZero)]))
+								.optional(),
+							flipX: z.literal(true).optional(),
+							flipY: z.literal(true).optional(),
+						})
+						.strict()
+						.refine(...nonEmpty)
+						.optional(),
```

Comment:
So, uh... is this accurate? Does this apply to _all_ `offset` properties, or even all `Vector2` types? tfw no documentation 😩